tests/src/3d/testqgs3drendering.cpp

```diff
@@ -485,6 +487,49 @@ void TestQgs3DRendering::testPhongShading()
   QGSVERIFYIMAGECHECK( "phong_shading", "phong_shading", img, QString(), 40, QSize( 0, 0 ), 2 );
 }
 
+void TestQgs3DRendering::testExtrudedPolygonsTexturedPhong()
+{
+  QgsPhongTexturedMaterialSettings materialSettings;
+  materialSettings.setAmbient( QColor( 26, 26, 26 ) );
+  materialSettings.setDiffuseTexture( QVariant::fromValue( testDataPath( "/sample_image.png" ) ) );
+  QgsPolygon3DSymbol *symbol3d = new QgsPolygon3DSymbol;
+  symbol3d->setMaterialSettings( materialSettings.clone() );
+  symbol3d->setExtrusionHeight( 10.f );
+  QgsVectorLayer3DRenderer *renderer3d = new QgsVectorLayer3DRenderer( symbol3d );
+  mLayerBuildings->setRenderer3D( renderer3d );
+
+  const QgsRectangle fullExtent = mLayerDtm->extent();
+
+  Qgs3DMapSettings *map = new Qgs3DMapSettings;
+  map->setCrs( mProject->crs() );
+  map->setExtent( fullExtent );
+  map->setLayers( QList<QgsMapLayer *>() << mLayerBuildings << mLayerRgb );
+  QgsPointLightSettings defaultLight;
+  defaultLight.setIntensity( 1.0 );
+  defaultLight.setPosition( QgsVector3D( 0, 1000, 0 ) );
+  map->setLightSources( { defaultLight.clone() } );
+
+  QgsFlatTerrainGenerator *flatTerrain = new QgsFlatTerrainGenerator;
+  flatTerrain->setCrs( map->crs() );
+  map->setTerrainGenerator( flatTerrain );
+
+  QgsOffscreen3DEngine engine;
+  Qgs3DMapScene *scene = new Qgs3DMapScene( *map, &engine );
+  engine.setRootEntity( scene );
+
+  scene->cameraController()->setLookingAtPoint( QgsVector3D( -60, 10, 360 ), 40, 45, 0 );
+
+  // When running the test on Travis, it would initially return empty rendered image.
+  // Capturing the initial image and throwing it away fixes that. Hopefully we will
+  // find a better fix in the future.
+  Qgs3DUtils::captureSceneImage( engine, scene );
+  QImage img = Qgs3DUtils::captureSceneImage( engine, scene );
+
+  delete scene;
+  delete map;
+  QGSVERIFYIMAGECHECK( "polygon3d_extrusion_textured_phong", "polygon3d_extrusion_textured_phong", img, QString(), 40, QSize( 0, 0 ), 2 );
```

Comment:
These test images look quite detailed, and I'm concerned they'll be tricky to make reliable. Could you increase the texture size to help alleviate this?

Reply:
The test (qt5 and qt6) works on my machine. The qt6 version works on the ci but not the qt5 version. 
I tried to zoom in a lot to verify that the texture is correct. I have updated the test to zoom out a bit.

Reply:
Not sure what happens. I'm positive this has nothing to do with the current PR. I got the same result on master. I have dropped this commit and created a separate PR to add the unit test: https://github.com/qgis/QGIS/pull/58188